tests/testthat/outputs/vars-select-bool-symbols.txt

```diff
@@ -1,10 +1,6 @@
 > vars_select(letters, starts_with("a") & z)
-Error: Can't use boolean operators with bare variables.
-x `z` is a bare variable.
-i Do you need `all_of(z)`?
+Error in eval_tidy(expr, context_mask): object 'z' not found
```

Comment:
Is this correct?

Reply:
Hmm I'm not sure anymore. Also I now realise the previous advice was not correct either (the `all_of(z)` wouldn't have worked, since `z` is a data variable here, not a context variable).

Here are the constraints:

- We disallow data lookup because users might be tempted to use `cyl & disp` where `cyl | disp` would be correct.
- We disallow context lookup of external vectors: user has to use `all_of()` in that case.
- We allow context lookup of predicate functions.

So what should be the error?

1. If we found an object that is not a function, we could throw a bad type error if the type is not right, or recommend `all_of()` if the type is adequate.

2. If we didn't find any object, the default "object not found" error seems ok.

Does that make sense? Currently we have error 2 but we're missing error 1.

Reply:
Hmmmm, I think it might be ok to allow bare variables here because (e.g.) `a | b | c` reads quite nicely and presumably `"a" | "b" | "c"` is ok? OTOH that introduces ambiguity if you have columns called `is.numeric()` so I think you're right.

I think it probably needs to be a specific error (as before) but with a hint that works — I guess the big question is what that should be. Maybe `starts_with("a") & c(z)`? But then that implies that `c(is.numeric)` would look for variable `is.numeric` which would be inconsistent with a bare `is.numeric`.

Or do we just say that variable names beat predicate names, and that's ok because it's sufficiently unusual to call a variable (e.g.) `is.numeric`?

Reply:
Perhaps a good principle to follow is that `select(a, b)` and `select(a | b)` should have exactly the same properties, otherwise it will be awkward to explain.

Also I realised that allowing full data + function scope in `a & b` can be useful too, for example `foo & is.numeric` means "select variable only if called foo and only if numeric".

The main question is what kind of error do we want when `a` is not a data frame variable nor a function. Should this be an object not found error, or an out-of-bounds error? One advantage of requiring wrapping in something like `where()` is that we solve the scoping ambiguity, and then having the OOB error is unambiguous as well.